colossalai/utils/model/lazy_init_context.py

```diff
@@ -187,71 +189,61 @@ def lazy_init_parameters(self, model: torch.nn.Module, device='cpu', call_back:
-                for element in arg:
-                    processed_element = _process_arg(element)
-                    rst_list.append(processed_element)
-                arg = rst_list
-            return arg
-
-        def _replace_meta_param_with_real_param(meta_param):
-            if meta_param.device != 'meta':
-                return meta_param
-            tensor_id = id(meta_param)
+        def _init_recurively(module):
+            for mod in module.children():
+                _init_recurively(mod)
+
+            for param in module.parameters(recurse=False):
+                _convert_to_real_tensor(param, module, is_buffer=False)
+
+            for buf in module.buffers(recurse=False):
+                _convert_to_real_tensor(buf, module, is_buffer=True)
+
+        def _convert_to_real_tensor(tensor, module, is_buffer):
+            if not tensor.is_meta:
+                return tensor
+            tensor_id = id(tensor)
             param_full_name = param_id_to_name[tensor_id]
-            real_param = torch.empty_like(meta_param, dtype=meta_param.dtype, device=device)
-            real_param = ColoParameter(real_param, requires_grad=meta_param.requires_grad)
+            real_tensor = torch.empty_like(tensor, dtype=tensor.dtype, device=device)
 
-            if '.' in param_full_name:
-                submodule_name, param_name = param_full_name.rsplit('.', 1)
-                submodule = model.get_submodule(submodule_name)
+            # look for initialization function
+            if tensor in self._intercepted_nn_init_func_cache:
+                init_func, args, kwargs = self._intercepted_nn_init_func_cache[tensor][-1]
             else:
-                submodule = model
-                param_name = param_full_name
-            setattr(submodule, param_name, real_param)
+                raise RuntimeError(f"{param_full_name} is not associated with a initialization function")
 
-            # execute call_back function on the materailized tensor
-            # this can where sharding comes in
-            if call_back:
-                call_back(real_param)
-            return real_param
+            # initialize
+            init_func(real_tensor, *args, **kwargs)
 
-        func, args, kwargs = self.module_rebuild_dict[model]
-        args = list(args)
+            # convert to colotensor
+            # if this is a parameter, we should convert it to coloparameter
+            # if this is a buffer, we should convert it to colotensor
+            if is_buffer:
+                real_tensor = ColoTensor.from_torch_tensor(real_tensor)
+            else:
+                real_tensor = ColoParameter.from_torch_tensor(real_tensor, requires_grad=tensor.requires_grad)
 
-        # check args for parameter replacement
-        for idx, arg in enumerate(args):
-            arg = _process_arg(arg)
-            args[idx] = arg
+            # convert to distribted mode
+            if hasattr(tensor, 'dist_spec'):
```

Comment:
isinstance(tensor, ColoTensor) ?

Reply:
`dist_spec` is annotated during the IR passes are executed, the IR pass will only set this attribute without effect since `tensor` is a meta tensor. The materialized `real_tensor` is guaranteed to be a ColoTensor and can redistribute.